text/3503-frontmatter.md

```diff
@@ -0,0 +1,512 @@
+
+Static site generators use a frontmatter syntax to embed metadata in markdown files:
+```md
+---
+name: My Blog Post
+---
+
+## Stuff happened
+
+Hello world!
+```
+
+We are carrying this concept over to Rust with a twist: using fence code blocks which
+will be familiar to Rust developers when documenting their code:
+````rust
+#!/usr/bin/env cargo
+```cargo
+[dependencies]
+clap = { version = "4.2", features = ["derive"] }
+```
+
+use clap::Parser;
+
+#[derive(Parser, Debug)]
+#[clap(version)]
+struct Args {
+    #[clap(short, long, help = "Path to config")]
+    config: Option<std::path::PathBuf>,
+}
+
+fn main() {
+    let args = Args::parse();
+    println!("{:?}", args);
+}
+````
+
+As we work to better understand how tool authors will want to use frontmatter, we are restricting it to just the `cargo` infostring.
+This means users will only be exposed to this within the concept of ["cargo script"](https://github.com/rust-lang/rfcs/pull/3502).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When parsing Rust code, after stripping the shebang (`#!`), rustc will strip a fenced code block:
+- Must be immediately at the top (after shebang stripping), meaning no blank lines
+- Opens with 3+ backticks and "cargo" followed by a newline
+  - We are intentionally supporting a subset of commonmark code fences in supporting backticks but not tilde code fences or indented fenced code blocks
+  - As we aren't supporting an arbitrarily nested file format (though may show up in one), we likely don't need the flexibility
+  - We are prioritizing on "one right way to do it" to make it easier to learn to write and to read a variety of files.
+- All content is ignored until the same number of backticks or more is found at the start of a line.
+  It is an error to have anything besides spaces and tabs between the backticks and the newline.
+  - Unlike commonmark, it is an error to not close the fenced code block seeing to detect problems earlier in the process seeing as the primary content is what comes after the fenced code block
+
+As cargo will be the first step in the process to parse this,
+the responsibility for high quality error messages will largely fall on cargo.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- A new concept for Rust syntax, adding to overall cognitive load
+- When sharing in markdown documents (e.g. GitHub issues), requires people escape markdown code fences with an extra backtick which they are likely not used to doing (or aware even exists)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Within this solution,
+we considered starting with only allowing this in the root `mod` (e.g. `main.rs`)
+but decided to allow it in any file mostly for ease of implementation.
+Like with Python, this allows any file in a package (with the correct deps and `mod`s) to be executed, allowing easier interacting experiences in verifying behavior.
```

Comment:
I don't think I'm quite seeing the concern.  They can be included in a published package much the same way you can include random data files.  We can't really do anything about random data files.  If its specifically about a `.rs` file that is both `mod`ed and a script, is there a reason we need to actively work to prevent that as its also unusable like data files?

Reply:
I guess the only potential concern I see is people being confused about the dependencies in the embedded manifest not being cared about at all. That might just be a theoretical concern though.

the comment above was meant a genuine open question, to figure out if anyone could think of any concerns like that. aside from the above I don't currently see any major issues with it.